### PR TITLE
Rework rundescriber and update to version2 

### PR DESCRIPTION
**Pull request by jenshnielsen**
Rework the rundescriber such only the newest version needs to exist in memory but can always be deserialzed from any older version and upgrade the newly written version of the rundescriber serialization to v2 that includes both `interdependences` and `interdependencies_` 

TODO

- [x] Update documentation
- [x] Ensure test coverage deserialization from any older version
- [x] Manual testing

### CODE REVIEW

File: qcodes/dataset/descriptions/versioning/rundescribertypes.py [hunk 1]
```diff
@@ -1,16 +1,35 @@
...
+This module defines the dict representation of the ``RunDescriber`` used for
+serialization and deserialization of the ``RunDescriber``.
+
+RunDescriber version log:
+
+- 0: The run_describer has a single attribute, interdependencies, which is an
+instance of InterDependencies (which contains ParamSpecs)
+- 1: The run_describer has a single attribute, interdependencies, which is an
+instance of InterDependencies_ (which contains ParamSpecBases)
+- 2: The run_describer has a two attribute, interdependencies, which is an
```
Comment (astafan8) [inline qcodes/dataset/descriptions/versioning/rundescribertypes.py:11]:
```suggestion
- 2: The run_describer has a two attributes: interdependencies, which is an
```

File: qcodes/dataset/descriptions/versioning/serialization.py [hunk 2]
```diff
@@ -1,90 +1,92 @@
 """
 The storage-facing module that handles serializations and deserializations
-into different versions of the top-level object, the RunDescriber.
+of the top-level object, the RunDescriber into and from different versions.
+
+Note that we require strict backwards and forwards compatibility such that
+the current RunDescriber must always be deserializable from any older or newer
+serialization.
+
+This means that a new version cannot delete omit previously included fields
```
Comment (astafan8) [inline qcodes/dataset/descriptions/versioning/serialization.py:9]:
```suggestion
This means that a new version cannot delete/omit previously included fields
```

**Review by astafan8 — COMMENTED** [review]
Looks great!

File: qcodes/dataset/descriptions/rundescriber.py [hunk 3]
```diff
@@ -25,33 +35,49 @@ def __init__(self, interdeps: InterDependencies_) -> None:
...
+            ser = cast(RunDescriberV0Dict, ser)
+            rundesc = cls(
+                old_to_new(InterDependencies._from_dict(ser['interdependencies']))
+            )
+        elif ser['version'] == 1:
+            ser = cast(RunDescriberV1Dict, ser)
+            rundesc = cls(
+                InterDependencies_._from_dict(ser['interdependencies'])
+            )
+        elif ser['version'] == 2:
```
Comment (jenshnielsen) [inline qcodes/dataset/descriptions/rundescriber.py:74]: This should probably be >=2 including a test

**Pull request merged**